File: 141. Linked List Cycle.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {boolean}
 */
var hasCycle = function(head) {
    if ( head == null || head.next == null ) {
        return false;
    }

    var fast = head.next;
    var slow = head;

    while ( fast != slow ) {
        if ( fast == null || fast.next == null ) {
            return false
        }
        fast = fast.next.next;
        slow = slow.next;
    }

    return true;
};

File: 141. Linked List Cycle.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution(object):
    def hasCycle(self, head):
        if head is None:
            return False

        slow = head
        fast = head.next

        while fast is not None and fast.next is not None:
            slow = slow.next
            fast = fast.next.next
            if slow == fast:
                return True
        return False
